File: Day13/day13.js
const fs = require('fs');
let map = [];
let foldingInstructions = [];

const readInputFileAndBuildMap = (test) => {
  let path = test ? 'testinput.txt' : 'input.txt';
  let data = fs.readFileSync(path, 'utf8');
  let lines = data.split('\n');
  let dots = lines.filter((line) => line.includes(',')).map((value) => value.split(',').map((val) => parseInt(val)));
  let largestX = 0;
  let largestY = 0;
  dots.forEach(element => {
    if (element[0] > largestX) {
      largestX = element[0];
    }
    if (element[1] > largestY) {
      largestY = element[1]
    }
  });

  for(let i = 0; i <= largestY; i++){
    map.push([])
    for(let j = 0; j <= largestX; j++) {
      map[i][j] = '.'
    }
  }


  dots.forEach(element => {
    map[element[1]][element[0]] = '#'
  })

  foldingInstructions = lines.filter((line) => line.includes('fold')).map((line) => line.split(' ')[2])
}


const horizontalFold = (foldIndex) => {
  let foldUp = false;

  if (foldIndex >= parseInt(map.length/2)) {
    foldUp = true
  }

  let numberOfIterations = 0;

  let startingHeight = foldUp ? foldIndex - 1 : 0
  let endingHeight = foldUp ? map.length : foldIndex
  for(let i = startingHeight; i < endingHeight; i++) {
    for(let j = 0; j < map[i].length; j++){
      if(map[i][j] == '#' && foldUp) {
        map[foldIndex - numberOfIterations + 1][j] = '#'
      } else if (map[i][j] == '#') {
        map[foldIndex + numberOfIterations + 1][j] = '#'
      }
    }
    numberOfIterations++;
  }

  map = map.filter((_x, index) => { 
    if (foldUp) {
      return index < foldIndex
    } else {
      return index > foldIndex
    }
  });
}

const verticalFold = (foldIndex) => {
  let foldRight = false;
  if (foldIndex <= parseInt(map[0].length/2)) {
    console.log('hey!')
    foldRight = true;
  }

  let startingWidth = foldRight ? 0 : foldIndex
  let endingWidth = foldRight ? foldIndex - 1: map[0].length

  for(let i = 0; i < map.length; i++) {
    let numberOfIterations = 0;
    for(let j = startingWidth; j < endingWidth; j++){
      if (map[i][j] == '#' && foldRight) {
        map[i][foldIndex + numberOfIterations] = '#'
      } else if(map[i][j] == '#') {
        map[i][foldIndex - numberOfIterations - 2] = '#'
      }
      numberOfIterations = numberOfIterations + 1;
    }
  }

  let newMap = []

  for(let i = 0; i < map.length; i++){
    newMap.push([]);
    if (foldRight) {
      for(let j = foldIndex; j < map[i].length; j++){
        newMap[i].push(map[i][j])
      }
    } else {
      for(let j = 0; j < foldIndex -1; j++){
        newMap[i].push(map[i][j])
      }
    }
  }

  map = newMap;
}

const runPart1 = () => {
  readInputFileAndBuildMap(true);
  foldingInstructions.forEach((instruction) => {
    let breakDown = instruction.split('')
    let axis = breakDown[0]
    let foldIndex = parseInt(breakDown[2]);
    printMap()
    console.log("Width ", map[0].length)
    console.log("Height ", map.length)
    console.log("about to fold ", instruction)
    console.log()
    console.log('*******')

    if (axis == 'y') {
      printHorizontalFold(foldIndex)
      horizontalFold(foldIndex)
    }
    if (axis == 'x') {
      printVerticalFold(foldIndex)
      verticalFold(foldIndex)
    }
    console.log()
    console.log()
  })

  printMap()

  let dotCount = 0;
  map.forEach((line) => {
    line.forEach((point) => {
      if (point == '#') {
        dotCount++;
      }
    })
  })

  console.log(dotCount)
};

const printMap = () => {
  for (let i = 0; i < map.length; i++) {
    console.log(map[i].join(""));
  }
}

const printHorizontalFold = (index) => {
  for (let i = 0; i < map.length; i++) {
    if (i == index) {
      console.log('-------------------------')
    } else {
      console.log(map[i].join(""));
    }
  }
}

const printVerticalFold = (index) => {
  for (let i = 0; i < map.length; i++) {
    let str = ''
    for(let j = 0; j < map[i].length; j++){
      if (j == index - 1) {
        str = str + '|'
      }
      else { 
        str = str + map[i][j]
      }
    }
    console.log(str);
  }
}

runPart1();

File: Day13/day13.rb
require 'pry'

class Day13
  attr_accessor :sheet, :instructions

  def initialize
    @sheet = []
    @instructions = []
  end

  def run_part_1(test_file = false )
    read_file(test_file)
    puts instructions
    instructions.each_with_index do |instruction, idx|
      split_up = instruction.split('=')
      axis = split_up[0]
      fold_point = split_up[1].to_i
      if (axis == 'y')
        print_horizontal_fold(fold_point)
        horizontal_fold(fold_point)
      else
        print_vertical_fold(fold_point)
        vertical_fold(fold_point)

      end
      puts
    end
    print_sheet
    puts count_hashtags
  end

  private

  def count_hashtags
    count = 0
    sheet.each do |line|
      line.each do |point|
        count = count + 1 if point == '#'
      end
    end
    count
  end

  def vertical_fold(fold_point)
    should_fold_left = fold_point >= (sheet[0].length/2)
    sheet.each_with_index do |line, line_idx|
      line.each_with_index do |point, point_idx|
        distance_from_fold_point = (point_idx - fold_point).abs
        next if fold_point == point_idx 

        if should_fold_left
          next if fold_point > point_idx
          if point == '#'
            sheet[line_idx][fold_point - distance_from_fold_point] = '#'
          end
        else
          next if fold_point < point_idx 
          if point == '#'
            sheet[line_idx][fold_point + distance_from_fold_point] = '#'
          end
        end
      end
    end


    new_sheet = []

    if should_fold_left
      sheet.each do |line|
        line.slice!(fold_point, line.length)
      end
    else
      sheet.each do |line|
        line.slice!(0, fold_point)
      end
    end

    sheet = new_sheet
    
  end

  def horizontal_fold(fold_point)
    should_fold_up = fold_point >= (sheet.length/2)

    sheet.each_with_index do |line, line_idx|
      distance_from_fold_point = (line_idx - fold_point).abs
      next if line_idx == fold_point

      if should_fold_up
        next if line_idx < fold_point
        line.each_with_index do |point, point_idx|
          if point == '#'
            sheet[fold_point - distance_from_fold_point][point_idx] = '#'
          end
        end
      else
        next if line_idx > fold_point
        line.each_with_index do |point, point_idx|
          if point == '#'
            sheet[fold_point + distance_from_fold_point][point_idx] = '#'
          end
        end
      end
    end

    if should_fold_up
      sheet.slice!(fold_point, sheet.length)
    else
      sheet.slice!(0, fold_point + 1)
    end
  end


  # HELPERS

  def print_sheet
    sheet.each do |line|
      puts line.join('')
    end
  end

  def print_horizontal_fold(fold_point)
    sheet.each_with_index do |line, index|
      if index == fold_point
        str = ''
        line.length.times do
          str = str + '-'
        end
        puts str
      else
        puts line.join('')
      end
    end
  end

  def print_vertical_fold(fold_point)
    str = ''
    sheet.each do |line|
      line.each_with_index do |point, index|
        if index == fold_point
          str = str + '|'
        else
          str = str + point
        end
      end
      puts str
      str = ''
    end
  end

  def read_file(test_file = false)
    file_path = test_file ? 'testinput.txt' : 'input.txt'
    file = File.open(file_path)
    reading_instructions = false
    max_x = 0;
    max_y = 0;

    file.readlines.each do |line|
      if reading_instructions
        instructions.push(line.split(' ').last)
      else
        if line.chomp == ''
          reading_instructions = true  
        else 
          coordinates = line.split(',').map(&:to_i)
          max_x = coordinates[0] if coordinates[0] > max_x
          max_y = coordinates[1] if coordinates[1] > max_y
        end
      end
    end

    (max_y + 1).times do 
      sheet.push([])
    end

    sheet.each do |empty_array|
      (max_x + 1).times do
        empty_array.push('.')
      end
    end

    file = File.open(file_path)
    file.readlines.each do |line|
      if line.chomp == ''
        break
      else
        coordinates = line.split(',').map(&:to_i)
        sheet[coordinates[1]][coordinates[0]]  = '#'
      end
    end
  end
end

Day13.new.run_part_1(false)